File: 300-longest-increasing-subsequence/solution.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        // dp[i] 表示以 nums[i] 这个数结尾的最长递增子序列的长度
        int[] dp = new int[nums.length];

        // base case：dp 数组全都初始化为 1
        Arrays.fill(dp, 1);

        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j])
                    dp[i] = Math.max(dp[i], dp[j] + 1);
            }
        }

        int res = 0;
        for (int i = 0; i < dp.length; i++) {
            res = Math.max(res, dp[i]);
        }
        return res;
    }
}

File: 300-longest-increasing-subsequence/solution.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = [1] * len(nums)  # basecase

        for i in range(len(nums)):
            for j in range(i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], 1 + dp[j])
        return max(dp)

        # top = [0] * len(nums)
        # piles = 0
        # for i, poker in enumerate(nums):
        #     left, right = 0, piles
        #     while (left < right):
        #         mid = (left + right) // 2
        #         if (top[mid] < poker):
        #             left = mid + 1
        #         elif (top[mid] > poker):
        #             right = mid
        #         else:
        #             right = mid
        #     if left == piles: piles += 1
        #     top[left] = poker
        # return piles
